docs/csharp/nullable-attributes.md

```diff
@@ -0,0 +1,276 @@
+
+## Specify Preconditions: `AllowNull` and `DisallowNull`
+
+Consider a read / write property that never returns `null` because it has a reasonable default value. Callers pass `null` to the set accessor when set it to that default value. For example, consider a messaging system that asks for a screen name in a chat room. If none is provided, the system generates a random name:
+
+```csharp
+public string ScreenName
+{
+   get { return screenName; }
+   set { screenName = value ?? GenerateRandomScreenName(); }
+}
+private string screenName;
+```
+
+When you compile the preceding code in a nullable oblivious context, everything is fine. Once you enable nullable reference types, the `ScreenName` property becomes a non-nullable reference. That's correct for the `get` accessor: it never returns `null`. Callers don't need to check the returned property for `null`. But now setting the property to `null` generates a warning. In order to continue to support this type of code, you add the <xref:System.Diagnostics.CodeAnalysis.AllowNull?displayProperty=nameWithType> attribute to the `set` accessor, as shown in the following code: 
+
+```csharp
+[AllowNull]
+public string ScreenName
+{
+   get { return screenName; }
+   set { screenName = value ?? GenerateRandomScreenName(); }
+}
+private string screenName = GenerateRandomScreenName();
+```
+
+You may need to add a `using` directive for `System.Diagnostics.CodeAnalysis` to use this and other attributes discussed in this article. The attribute is applied to the property, not the `set` accessor. The `AllowNull` attribute specifies *pre-conditions*, and only applies to inputs. The `get` accessor has a return value, but no input arguments. Therefore, the `AllowNull` attribute only applies to the `set` accessor.
+
+The preceding example demonstrates what to look for when adding the `AllowNull` attribute on an argument:
+
+1. The general contract for that variable is that it shouldn't be `null`, so you want a non-nullable reference type.
+1. There are scenarios for the input variable to be `null`, though they aren't the most common usage.
+
+Most often you'll need this attribute for properties, or `in` `out` and `ref` arguments. The `AllowNull` attribute is the best choice when a variable is typically non-null, but you need to allow `null` as a precondition.
+
+Contrast that with scenarios for using `DisallowNull`: You use this attribute to specify that an input variable of a nullable type shouldn't be `null`. Consider a property where `null` is the default value, but clients can only set it to a non-null value. Consider the following code:
+
+```csharp
+public string ReviewComment // Comments can be set, but not cleared.
+{
+    get { return _comment;}
+    set
+    {
+        if (value == null) throw new ArgumentNullException(nameof(value), "Cannot set to null");
+        _comment = null;
```

Comment:
I think you may be setting it to `null` as a way to let the `_comment` variable become `null`, but that really is just confusing as it's not realistic and I'm sure you'll get comments on it. What about just setting the initial value to null or providing a simple `public void Reset() => _comment = null` method?

```suggestion
        _comment = value;
```